From 02684a2e215a96a0b82b5a644fc2c0d0a4864fdc Mon Sep 17 00:00:00 2001
From: Lukas Czerner <lczerner@redhat.com>
Date: Thu, 23 Oct 2014 09:55:11 -0400
Subject: [fs] ext4: fix unjournalled bg descriptor while initializing inode bitmap

Message-id: <1414058125-4183-158-git-send-email-lczerner@redhat.com>
Patchwork-id: 98226
O-Subject: [RHEL 7.1 PATCH v2 157/171] ext4: fix unjournalled bg descriptor while initializing inode bitmap
Bugzilla: 1150139
RH-Acked-by: Eric Sandeen <sandeen@redhat.com>

From: Theodore Ts'o <tytso@mit.edu>

BZ 1150139
https://bugzilla.redhat.com/show_bug.cgi?id=1150139

BREW 8142710
https://brewweb.devel.redhat.com/taskinfo?taskID=8142710

Upstream commit 61c219f5814277ecb71d64cb30297028d6665979

 The first time that we allocate from an uninitialized inode allocation
 bitmap, if the block allocation bitmap is also uninitalized, we need
 to get write access to the block group descriptor before we start
 modifying the block group descriptor flags and updating the free block
 count, etc.  Otherwise, there is the potential of a bad journal
 checksum (if journal checksums are enabled), and of the file system
 becoming inconsistent if we crash at exactly the wrong time.

 Signed-off-by: Theodore Ts'o <tytso@mit.edu>
 Cc: stable@vger.kernel.org
 Signed-off-by: Lukas Czerner <lczerner@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/ext4/ialloc.c b/fs/ext4/ialloc.c
index 0ee59a6..64bb32f 100644
--- a/fs/ext4/ialloc.c
+++ b/fs/ext4/ialloc.c
@@ -851,6 +851,13 @@ got:
   goto out;
  }
 
+ BUFFER_TRACE(group_desc_bh, "get_write_access");
+ err = ext4_journal_get_write_access(handle, group_desc_bh);
+ if (err) {
+  ext4_std_error(sb, err);
+  goto out;
+ }
+
  /* We may have to initialize the block bitmap if it isn't already */
  if (ext4_has_group_desc_csum(sb) &&
      gdp->bg_flags & cpu_to_le16(EXT4_BG_BLOCK_UNINIT)) {
@@ -887,13 +894,6 @@ got:
   }
  }
 
- BUFFER_TRACE(group_desc_bh, "get_write_access");
- err = ext4_journal_get_write_access(handle, group_desc_bh);
- if (err) {
-  ext4_std_error(sb, err);
-  goto out;
- }
-
  /* Update the relevant bg descriptor fields */
  if (ext4_has_group_desc_csum(sb)) {
   int free;
-- 
1.7.1